### PR TITLE
Add SecretControlStore.lockProjectSecret

### PR DESCRIPTION
**Pull request by frsyuki**
Some secret operations such as refreshing OAuth token need inter-sever
mutual exclusion. This new method provides locking without depending on
another external components.

### CODE REVIEW

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSecretControlStore.java [hunk 1]
```diff
@@ -50,15 +51,81 @@ public void setProjectSecret(int projectId, String scope, String key, String val
...
+
+            // TODO this logic is copied from DatabaseSecretStore.getSecret.
+            Optional<String> value;
+            if (secret == null) {
+                value = Optional.absent();
+            }
+            else {
+                // TODO: look up crypto engine using name
+                if (!crypto.getName().equals(secret.engine)) {
+                    throw new AssertionError("Crypto engine mismatch");
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSecretControlStore.java]: Adding the values of `crypt.getName()` and `secret.engine` into the message would be better.

File: digdag-core/src/test/java/io/digdag/core/database/DatabaseSecretStoreTest.java [hunk 2]
```diff
@@ -82,6 +82,18 @@ public void singleSecret()
...
 
     @Test
+    public void lockSecret()
+            throws Exception
+    {
+        Optional<String> absent = factory.begin(() -> secretControlStore.lockProjectSecret(projectId, SecretScopes.PROJECT, KEY1, (control, value) -> value));
+        assertThat(absent, is(Optional.absent()));
+
+        factory.begin(() -> secretControlStore.setProjectSecret(projectId, SecretScopes.PROJECT, KEY1, VALUE1));
+        Optional<String> present = factory.begin(() -> secretControlStore.lockProjectSecret(projectId, SecretScopes.PROJECT, KEY1, (control, value) -> value));
```
Comment (komamitsu) [inline digdag-core/src/test/java/io/digdag/core/database/DatabaseSecretStoreTest.java]: We'd better have a test that confirms if another thread is blocked with the method

**Comment by komamitsu** [comment]
@frsyuki I left a comment https://github.com/treasure-data/digdag/pull/526/files#r108588806 and another minor comment. Can you look at the first one?

File: digdag-core/src/test/java/io/digdag/core/database/DatabaseSecretStoreTest.java (hunk 2 above)
Comment (frsyuki) [inline digdag-core/src/test/java/io/digdag/core/database/DatabaseSecretStoreTest.java]: Good point. I added a test.

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseSecretControlStore.java (hunk 1 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseSecretControlStore.java]: yes, true. I don't know why here uses AssertionError though...

**Comment by komamitsu** [comment]
👍

**Pull request merged**